src/parser/parser.cc

```diff
@@ -0,0 +1,2496 @@
+    {
+      // select <- postfix '.' (ident / symbol)
+      if (!has(TokenKind::Dot))
+        return Skip;
+
+      auto sel = std::make_shared<Select>();
+      sel->location = previous.location;
+      sel->expr = expr;
+      expr = sel;
+
+      if (has(TokenKind::Ident) || has(TokenKind::Symbol))
+      {
+        sel->member = previous.location;
+        return Success;
+      }
+
+      error() << loc() << "Expected an identifier or a symbol" << line();
+      return Error;
+    }
+
+    Result optstaticselect(Node<Expr>& expr)
+    {
+      // staticselect <- postfix ('::' (ident / symbol) typeargs?)+
+      if (!has(TokenKind::DoubleColon))
+        return Skip;
+
+      auto sel = std::make_shared<StaticSelect>();
+      sel->location = previous.location;
+      sel->expr = expr;
+      expr = sel;
+
+      Result r = Success;
+
+      while (has(TokenKind::DoubleColon))
+      {
+        if (!has(TokenKind::Ident) && !has(TokenKind::Symbol))
+        {
+          error() << loc() << "Expected an identifier or symbol" << line();
+          r = Error;
+          break;
+        }
+
+        // Use the location of the last ident or symbol.
+        sel->location = previous.location;
+
+        auto name = std::make_shared<TypeName>();
+        name->location = previous.location;
+        sel->typenames.push_back(name);
+
+        if (opttypeargs(name->typeargs) == Error)
+          r = Error;
+      }
+
+      return r;
+    }
+
+    Result opttypeargs(List<Type>& typeargs)
+    {
+      // typeargs <- '[' (expr (',' expr)*)?) ']'
+      if (!has(TokenKind::LSquare))
+        return Skip;
+
+      Result r = Success;
+
+      do
+      {
+        Node<Type> arg;
+
+        if (typeexpr(arg) != Success)
+        {
+          error() << loc() << "Expected a type argument" << line();
+          restart_before({TokenKind::Comma, TokenKind::RSquare});
+          r = Error;
+        }
+
+        typeargs.push_back(arg);
+      } while (has(TokenKind::Comma));
+
+      if (!has(TokenKind::RSquare))
+      {
+        error() << loc() << "Expected , or ]" << line();
+        r = Error;
+      }
+
+      return r;
+    }
+
+    Result optspecialise(Node<Expr>& expr)
+    {
+      // specialise <- postfix typeargs
+      if (!peek(TokenKind::LSquare))
+        return Skip;
+
+      rewind();
+      auto spec = std::make_shared<Specialise>();
+      spec->location = previous.location;
+      spec->expr = expr;
+      expr = spec;
+
+      if (opttypeargs(spec->typeargs) != Success)
+        return Error;
+
+      return Success;
+    }
+
+    Result onepostfix(Node<Expr>& expr)
+    {
+      Result r;
+
+      if ((r = optstaticselect(expr)) != Skip)
+        return r;
+
+      if ((r = optselect(expr)) != Skip)
+        return r;
+
+      if ((r = optspecialise(expr)) != Skip)
+        return r;
+
+      return Skip;
+    }
+
+    Result optpostorblock(Node<Expr>& expr)
+    {
+      // postfix <- atom ('.' (ident / symbol) / typeargs)*
```

Comment:
orblock in name, doesn't seem to be reflected in the comment?

Reply:
Improved comment!